lib/posix/pthread.c

```diff
@@ -254,31 +407,32 @@ int pthread_setcancelstate(int state, int *oldstate)
  */
 int pthread_cancel(pthread_t pthread)
 {
-	struct posix_thread *thread = to_posix_thread(pthread);
+	void *retval;
+	int detachstate;
 	int cancel_state;
-	k_spinlock_key_t cancel_key;
+	k_spinlock_key_t key;
+	struct posix_thread *t;
 
-	if ((thread == NULL) || (thread->state == PTHREAD_TERMINATED)) {
+	t = to_posix_thread(pthread);
+	if (t == NULL) {
 		return ESRCH;
 	}
 
-	cancel_key = k_spin_lock(&thread->cancel_lock);
-	thread->cancel_pending = 1;
-	cancel_state = thread->cancel_state;
-	k_spin_unlock(&thread->cancel_lock, cancel_key);
-
+	key = k_spin_lock(&q_lock);
+	t->cancel_pending = true;
+	cancel_state = t->cancel_state;
+	detachstate = t->detachstate;
 	if (cancel_state == PTHREAD_CANCEL_ENABLE) {
```

Comment:
I think the whole retval management is incorrect here -- if the thread is canceled, then the retval seen through `pthread_join()` should be `PTHREAD_CANCELED`. So, I think this chunk of code can simply be deleted and the value `PTHREAD_CANCELED` be passed unconditionally to `posix_thread_finalize()`. That code should make sure the thread is not already terminated before assigning the new retval to `t->retval` so that calls to `pthread_cancel()` don't override a simultaneous call to `pthread_exit()`.